tests/PathMappingRepositoryTest.php

```diff
@@ -0,0 +1,405 @@
+<?php
+
+/*
+ * This file is part of the puli/repository package.
+ *
+ * (c) Bernhard Schussek <bschussek@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Repository\Tests;
+
+use Puli\Repository\Api\EditableRepository;
+use Puli\Repository\Api\Resource\Resource;
+use Puli\Repository\PathMappingRepository;
+use Puli\Repository\Resource\DirectoryResource;
+use Puli\Repository\Resource\FileResource;
+use Puli\Repository\Tests\Resource\TestFilesystemDirectory;
+use Puli\Repository\Tests\Resource\TestFilesystemFile;
+use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\KeyValueStore\ArrayStore;
+
+/**
+ * @since  1.0
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class PathMappingRepositoryTest extends AbstractEditableRepositoryTest
+{
+    /**
+     * @var ArrayStore
+     */
+    protected $store;
+
+    /**
+     * @var PathMappingRepository
+     */
+    protected $repo;
+
+    /**
+     * Temporary directory for test files.
+     *
+     * @var string
+     */
+    protected $tempDir;
+
+    /**
+     * Counter to avoid collisions during tests on directories.
+     *
+     * @var int
+     */
+    protected static $createdDirectories = 0;
+
+    /**
+     * Counter to avoid collisions during tests on files.
+     *
+     * @var int
+     */
+    protected static $createdFiles = 0;
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->tempDir = sys_get_temp_dir().'/puli-repository/PathMappingRepositoryTest'.md5(uniqid(time(), true));
+        mkdir($this->tempDir, 0777, true);
+
+        $this->store = new ArrayStore();
+        $this->repo = new PathMappingRepository($this->store);
+    }
+
+    protected function tearDown()
+    {
+        parent::tearDown();
+
+        $filesystem = new Filesystem();
+        $filesystem->remove($this->tempDir);
+    }
+
+    protected function createPrefilledRepository(Resource $root)
+    {
+        $repo = new PathMappingRepository(new ArrayStore());
+        $repo->add('/', $root);
+
+        return $repo;
+    }
+
+    protected function createWriteRepository()
+    {
+        return new PathMappingRepository(new ArrayStore());
+    }
+
+    protected function createReadRepository(EditableRepository $writeRepo)
+    {
+        return $writeRepo;
+    }
+
+    protected function createFile($path = null, $body = TestFilesystemFile::BODY)
+    {
+        return new TestFilesystemFile($path, $body);
+    }
+
+    protected function createDirectory($path = null, array $children = array())
+    {
+        return new TestFilesystemDirectory($path, $children);
+    }
+
+    protected function buildStructure(Resource $root)
```

Comment:
I tried to merge it but I can't really know how to build a structure from the inside of `createDirectory` for instance. I mean, for instance, in the following coded, I'm not able to know that the first createDirectory should build the file structure where the inside one should not:

``` php
$this->createDirectory('', array(
    $this->createDirectory('/sub1', array(
        $this->createFile('/file1', 'original 1'),
    )),
    $this->createDirectory('/sub2', array(
        $this->createFile('/file2', 'original 2'),
    )),
))
```